proposals/crd-support.md

```diff
@@ -0,0 +1,52 @@
+
+## Motivation
+The current ways to handle CRDs in Helm have issues and these are known to have caused real problems for users (more details in [#4863](https://github.com/helm/helm/issues/4863)). The goal here is to handle CRDs in a way that is intuitive and flexible for chart developers, as well as easy to use and safe for helm users. The goal is that this could be introduced in Helm 2 as well as make it into Helm 3.
+
+## Goals
+* The solution should not be dependent on versioned CRDs. Even though these are in development, a solution needs to work with existing CRDs. But we would like a solution that can work with/take advantage of versioned CRDs when they become available.
+* It should support existing installed charts that use CRDs, regardless of which of the existing ways to install CRDs was used.
+* CRDs should never be deleted unless the user has explicitely asked for it. Since CRDs are global in a Kubernetes cluster, we can never know for sure that it is no longer in use.
+* CRDs should never be updated unless the user has explicitely asked for it. Updating CRDs could render them unusable for some of the services/applications that depend on them.
+* Mixing CRDs and CRs of the type in the same chart should be supported.
+
+
+## Implementation
+This proposal takes a very conservative approach on any changes to CRDs, but the hope is that it will not be to inconvenient. If CRDs are either only used by a single chart, or is used by many charts but rarely updated (like App CRD), updates should usually be possible without too much inconvenience for the user.
+
+CRDs need to be handled correctly through all the phases of a Helm release:
+
+### Install
+* CRDs should be the first resource processed during installation. CRDs must be installed into the cluster before any instances of the CRD.
+* If a chart contains one or more CRDs, the helm install process should for each of them, check if the CRD is already installed. 
+  * If it is, Helm should check if the installed CRD is identical to the one contained in the chart. 
+    * If it is, the install process can skip installing the CRD and just continue.
+    * If it is different, the install process should fail, unless the user has explicitly confirmed through a command line flag that the CRD in the chart should overwrite the existing one.
+  * If it does not already exist, the CRD will be installed.
+* All CRDs in the chart will be tracked as part of a release when the chart is installed
+  * If the CRD was already installed, then the existing CRD will be adopted into the release.
+
+### Delete
+* When a chart is deleted, all CRDs will be orphaned, unless the user has explicitly through a command line flag decided that it should be deleted.
```

Comment:
Just observing that the delete flag will need to be specified per CRD.

Reply:
Correct, it would require the user to specify which CRDs should be deleted in the situation where a chart contains more than one. The same applies to updates to CRDs.

Reply:
Should there be a CLI flag to delete all CRDs and not just those specified? It would be a nice shortcut for testing.

Reply:
I am NOT a fan of having to specify individual CRD names to delete. It's tedious, and requires the user to know very deeply what is happening behind the scenes. Considering that there are Kubernetes tools that install more than a dozen CRDs, this can get pretty onerous.

Reply:
The thought here was that this would provide a way for users to delete CRDs with a chart, but that the default would be to orphan the CRDs which is the same as the current crd_install hook. The issue here is that there are almost no scenarios when it will be safe to delete a CRD, so deleting any CRD will require some kind of confirmation from the user.
I guess it would be possible to find some way for Helm to count the number of releases that use a specific CRD, and then know that the CRD can be deleted when the last release is deleted. But it seems pretty complicated and it also wouldn't be able to track any usage of a CRD that doesn't go through Helm.

Reply:
CRDs are an admin task. You need high level access to even install them. Something many Helm users may not have as corporate environments are more locked down. The person installing an app/CRD and maybe later uninstalling it may likely not have cluster wide context.

I wonder if an admin for the cluster would have enough context to the workloads running in the cluster to know when to delete a CRD.

Is this lack of context, especially in large clusters, a CRD shortcoming?

Reply:
Agree with @mortent that automating deletes using a Garbage collection like counting approach might be possible, but seems too complicated.

@mattfarina Isn't the 'context' that you refer to essentially the same thing as the various helm releases in which CRD is being used (from @mortent's comment). 
How to use that context (whether programmatically by 'helm delete' or by admin before calling 'helm delete'), that seems to be the main problem. I would not categorize that being a shortcoming of CRD.